### PR TITLE
Only run event systems if they have tangible work to do

### PR DESCRIPTION
**Pull request by james7132**
# Objective
Scheduling low cost systems has significant overhead due to task pool contention and the extra machinery to schedule and run them. Event update systems are the prime example of a low cost system, requiring a guaranteed O(1) operation, and there are a *lot* of them.

## Solution
Add a run condition to every event system so they only run when there is an event in either of it's two internal Vecs.

---

## Changelog
Changed: Event update systems will not run if there are no events to process.

## Migration Guide
`Events<T>::update_system` has been split off from the the type and can be found at `bevy_ecs::event::event_update_system`.

### CODE REVIEW

File: crates/bevy_ecs/src/event.rs [hunk 1]
```diff
@@ -552,6 +552,12 @@ impl<E: Event> Events<E> {
         events.update();
     }
 
+    /// A [`Condition`] that checks if the event's [`update_system`](Self::update_system)
+    /// needs to run or not.
+    pub fn update_condition(events: Res<Self>) -> bool {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/event.rs:557]: I think this would read more clearly as `need_updating`?

Comment (james7132) [reply crates/bevy_ecs/src/event.rs:557]: That might work, but the fact this takes a `Res<Self>` doesn't conform with that kind of use. Perhaps we could emulate the `update`/`update_system` split for this, though I'm not a huge fan of putting the system on the type itself.

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/event.rs:557]: > though I'm not a huge fan of putting the system on the type itself.

Agreed, that pattern here has always struck me as pretty odd and hard to read.

Comment (aevyrie) [reply crates/bevy_ecs/src/event.rs:557]: > this takes a `Res<Self>`

Maybe `has_updates`? Would line up with `&self -> bool`

```rs
.run_if(Events::<T>::has_updates),
```

Comment (james7132) [reply crates/bevy_ecs/src/event.rs]: I extracted both of these so they exist separately from the `Events<T>` type.

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_ecs/src/event.rs [hunk 2]
```diff
@@ -754,6 +749,17 @@ impl<'a, E: Event> ExactSizeIterator for EventIteratorWithId<'a, E> {
     }
 }
 
+/// A system that calls [`Events::update`] once per frame.
+pub fn event_update_system<T: Event>(mut events: ResMut<Events<T>>) {
+    events.update();
+}
+
+/// A run condition that checks if the event's [`update_system`]
```
Comment (IceSentry) [inline crates/bevy_ecs/src/event.rs:757]:
```suggestion
/// A run condition that checks if the event's [`event_update_system`]
```

**Review by IceSentry — APPROVED** [review]
Small nit but other than that LGTM

**Review by mockersf — APPROVED** [review]

**Pull request merged**